components/nexus-repository/src/main/java/org/sonatype/nexus/repository/util/SimpleStateGuard.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Sonatype Nexus (TM) Open Source Version
+ * Copyright (c) 2007-2015 Sonatype, Inc.
+ * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+
+package org.sonatype.nexus.repository.util;
+
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+
+import javax.annotation.Nullable;
+
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Throwables;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Simple state guard.
+ *
+ * @since 3.0
+ */
+public class SimpleStateGuard<T extends Enum>
+    extends ComponentSupport
+{
+  private final Lock lock;
+
+  @Nullable
+  private final T failure;
+
+  private T current;
+
+  public SimpleStateGuard(final Lock lock,
+                          final T initial,
+                          final @Nullable T failure)
+  {
+    this.lock = checkNotNull(lock);
+    this.current = checkNotNull(initial);
+    this.failure = failure;
+  }
+
+  private void lock() {
```

Comment:
Do we expect more threads checking guards vs performing transitions? If so then a reader-writer lock might be something to look into.